presto-verifier/src/main/java/com/facebook/presto/verifier/retry/RetryDriver.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.verifier.retry;
+
+import com.facebook.presto.verifier.framework.QueryException;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Predicate;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.Math.pow;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class RetryDriver
+{
+    private static final Logger log = Logger.get(RetryDriver.class);
+
+    private final int maxAttempts;
+    private final Duration minBackoffDelay;
+    private final Duration maxBackoffDelay;
+    private final double scaleFactor;
+    private final Predicate<QueryException> retryPredicate;
+
+    public RetryDriver(
+            RetryConfig config,
+            Predicate<QueryException> retryPredicate)
+    {
+        this.maxAttempts = config.getMaxAttempts();
+        this.minBackoffDelay = requireNonNull(config.getMinBackoffDelay(), "minBackoffDelay is null");
+        this.maxBackoffDelay = requireNonNull(config.getMaxBackoffDelay(), "maxBackoffDelay is null");
+        this.scaleFactor = config.getScaleFactor();
+        this.retryPredicate = requireNonNull(retryPredicate, "retryPredicate is null");
+    }
+
+    public <V> V run(String callableName, RetryOperation<V> operation)
+    {
+        int attempt = 1;
+        while (true) {
+            try {
+                return operation.run();
+            }
+            catch (QueryException qe) {
+                if (attempt >= maxAttempts || !retryPredicate.test(qe)) {
+                    throw qe;
```

Comment:
In my opinion, we need a way to record failed runs. Specifically - if some changes increase the inter-worker communication failures, we want to record that information - and surface it up. So, I want to able to see on the UI, queries are mostly succeeding in their 3rd try as compared to in the first one. Also, which errors happened will be useful information to have.